src/main/kotlin/org/rust/lang/core/lexer/RustStringLiteralLexer.kt

```diff
@@ -0,0 +1,129 @@
+package org.rust.lang.core.lexer
+
+import com.intellij.openapi.util.text.StringUtil
+import com.intellij.psi.StringEscapesTokenTypes.*
+import com.intellij.psi.tree.IElementType
+
+const val BYTE_ESCAPE_LENGTH = "\\x00".length
+const val UNICODE_ESCAPE_MIN_LENGTH = "\\u{0}".length
+const val UNICODE_ESCAPE_MAX_LENGTH = "\\u{000000}".length
+
+/**
+ * Performs lexical analysis of Rust byte/char/string/byte string literals using Rust character escaping rules.
+ */
+public class RustStringLiteralLexer(val defaultToken: IElementType,
+                                    val byte: Boolean = false,
+                                    val unicode: Boolean = false,
+                                    val eol: Boolean = false) : LexerBaseKt() {
+    override fun start(buffer: CharSequence, startOffset: Int, endOffset: Int, initialState: Int) {
+        bufferSequence = buffer
+        bufferEnd = endOffset
+        state = initialState
+
+        tokenStart = startOffset
+        tokenEnd = locateToken(tokenStart)
+    }
+
+    override fun advance() {
+        tokenStart = tokenEnd
+        tokenEnd = locateToken(tokenStart)
+    }
+
+    override fun getTokenType(): IElementType? {
+        // We're at the end of the string token => finish lexing
+        if (tokenStart >= tokenEnd) {
+            return null
+        }
+
+        // We're not inside escape sequence
+        if (bufferSequence[tokenStart] != '\\') {
+            return defaultToken
+        }
+
+        // \ is at the end of the string token
+        if (tokenStart + 1 >= tokenEnd) {
+            return INVALID_CHARACTER_ESCAPE_TOKEN
+        }
+
+        return when (bufferSequence[tokenStart + 1]) {
+            'u' ->
+                if (!unicode) {
+                    INVALID_CHARACTER_ESCAPE_TOKEN
+                } else if (isValidUnicodeEscape(tokenStart, tokenEnd)) {
+                    VALID_STRING_ESCAPE_TOKEN
+                } else {
+                    INVALID_UNICODE_ESCAPE_TOKEN
+                }
```

Comment:
No, it's not super important (micro-optimisation: check `!unicode` before `isValidUnicodeEscape`). Asked just out of curiosity :smile: